src/machine/parsed_results.rs

```diff
@@ -141,6 +79,206 @@ pub enum Value {
+        let mut conj_term = Value::Structure(",".into(), vec![beflast, last]);
+
+        for val in conj_rev {
+            conj_term = Value::Structure(",".into(), vec![val.clone(), conj_term]);
+        }
+
+        conj_term
+    }
+
+    pub fn disjunction(disj: &[Value]) -> Value {
+        if disj.is_empty() {
+            panic!("can't create empty disjunction");
+        } else if disj.len() == 1 {
+            return disj[0].clone();
+        }
+
+        let mut disj_rev = disj.iter().rev();
+
+        let last = disj_rev.next().unwrap().clone();
+        let beflast = disj_rev.next().unwrap().clone();
+        let mut disj_term = Value::Structure(";".into(), vec![beflast, last]);
+
+        for val in disj_rev {
+            disj_term = Value::Structure(";".into(), vec![val.clone(), disj_term]);
+        }
+
+        disj_term
+    }
+}
+
+impl From<&QueryResolutionLine> for Value {
+    fn from(qrl: &QueryResolutionLine) -> Self {
+        match qrl {
+            QueryResolutionLine::True => Value::Atom("true".into()),
+            QueryResolutionLine::False => Value::Atom("false".into()),
+            QueryResolutionLine::Match(m) => Value::conjunction(
+                &m.bindings
+                    .iter()
+                    .map(|(k, v)| {
+                        Value::Structure("=".into(), vec![Value::Var(k.clone()), v.clone()])
+                    })
+                    .collect::<Vec<_>>(),
+            ),
+        }
+    }
+}
+
+impl From<&QueryResolution> for Value {
+    fn from(qr: &QueryResolution) -> Self {
+        match qr {
+            QueryResolution::True => Value::Atom("true".into()),
+            QueryResolution::False => Value::Atom("false".into()),
+            QueryResolution::Matches(matches) => Value::disjunction(
+                &matches
+                    .iter()
+                    .map(|m| {
+                        Value::conjunction(
+                            &m.bindings
+                                .iter()
+                                .map(|(k, v)| {
+                                    Value::Structure(
+                                        "=".into(),
+                                        vec![Value::Var(k.clone()), v.clone()],
+                                    )
+                                })
+                                .collect::<Vec<_>>(),
+                        )
+                    })
+                    .collect::<Vec<_>>(),
+            ),
+        }
+    }
+}
+
+struct ValueRationalInner {
+    up: IBig,
+    down: UBig,
+}
+
+impl Serialize for ValueRationalInner {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(2))?;
+        match self.up.to_i64() {
+            Some(small) => seq.serialize_element(&small)?,
+            None => seq.serialize_element(&self.up.to_string())?,
+        }
+        match self.down.to_i64() {
```

Comment:
Given that `down` is unsigned and the PR description mentions 

> For integers that don't fit in an i64 (or in an u64 in the case of the denominator of a rational) they are represented as strings.

I would expect this to use `to_u64` not `to_i64` (assuming that exists).

```suggestion
        match self.down.to_u64() {
```